Add conditions to SpecialResource status

pmtk opened the pull request:
PR changes SpecialResource's Status by removing State (`string`) and using Conditions.
Three Conditions are expected to be used:
- `Ready` - SpecialResource is operational, ie. the recipe was successfully reconciled,
- `Progressing` - handling SpecialResource's recipe is in progress,
- `Errored` - an error occurred during handling SpecialResource or its recipe. It might be a short-lived error that will be might very quickly change into Progressing, or an error that might require user intervention (such as incorrect recipe).

Only one Condition is expected to be `true` at given moment, while the others should be `false`, ie. they're mutually exclusive.
Conditions' statuses (`true` or `false`; `unknown` is unused) are presented as additional columns when executing `{oc,kubectl} get specialresource`.

Current condition changes (lifecycle of SR within single reconciliation):
- `controllers/specialresource.go` `SpecialResourcesReconcile()`
  - SR to reconcile is found in the API
    Cond: Progressing, Reason: Progressing
  - SR is marked for deletion
    Cond: Progressing, Reason: MarkedForDeletion
  - loading chart fails for either SR or dependency
    Cond: Errored, Reason: {Dependency,}ChartFailure
  - information about dependency fails to be stored
    Cond: Errored, Reason: FailedToStoreDependencyInfo
  - attempt to create SR for dependency fails
    Cond: Errored, Reason: FailedToCreateDependencySR
- `controllers/resources.go` `ReconcileChartStates()`
  - ctrl starts to execute state yaml
    Cond: Progressing, Reason: HandlingState, Message: "Working on: templates/0000_buildconfig.yml"
- again `controllers/specialresource.go` `SpecialResourcesReconcile()`
  - attempt to deploy dependency or parent chart fails
    Cond: Errored, Reason: FailedToDeploy{Dependency,}Chart
  - reconciliation succeeds
    Cond: Ready, Reason: Success

Consequences of current implementation:
- Errors are very-short-lived. In case of reconciling just one SR at the time, `Errored` conditions might last just a couple of seconds. This means that SRs will have following Conditions during "happy scenario":
  - `Progressing + Progressing`
  - `Progressing + HandlingState + Msg: "Working on: templates/0000.yaml"`
  - `Errored + FailedToDeploy + Msg: "timed out waiting for..."` for couple seconds
  - `Progressing + Progressing` as new reconciliation starts
  - above will repeat couple of times
  - `Ready + Success`
- Another consequence of flow presented above is that re-reconciliation resets a status to `Progressing + Progressing` instead of `Progressing + HandlingState + Msg: "Working on: ..."`

pmtk committed:
Use metav1.Condition

pmtk committed:
Add operations on SR's Conditions

pmtk committed:
Introduce error to SetAs*. Regenerate mock.

pmtk committed:
Unit tests

pmtk committed:
Move Progressing's "Handling State" earlier

pmtk committed:
Fix linter

pmtk committed:
Remove old UpdateWithState

pmtk committed:
Add doc to functions

pmtk committed:
Align name of error used in StatusUpdater calls

pmtk committed:
Correct Errored status description

pmtk committed:
Make Condition's Reasons a const string

pmtk commented:
/cc @yevgeny-shnaidman @pacevedom @qbarrand

pmtk commented:
/ok-to-test

qbarrand commented on controllers/resources.go:
```diff
@@ -147,6 +148,9 @@ func ReconcileChartStates(ctx context.Context, r *SpecialResourceReconciler) err
 	for _, stateYAML := range stateYAMLS {
 
 		log.Info("Executing", "State", stateYAML.Name)
+		if suErr := r.StatusUpdater.SetAsProgressing(ctx, &r.specialresource, s.HandlingState, fmt.Sprintf("Working on: %s", stateYAML.Name)); suErr != nil {
```
Why `suErr`?

pmtk replied:
Because of consistency between this line and similar ones in `controllers/specialresource.go`

qbarrand commented on internal/controllers/state/statusupdater.go:
```diff
@@ -31,40 +53,71 @@ func NewStatusUpdater(kubeClient clients.ClientsInterface) StatusUpdater {
 		isMarkedToBeDeleted := sr.GetDeletionTimestamp() != nil
 		if isMarkedToBeDeleted {
-			return
+			return errors.New("Status won't be updated because object is marked for deletion")
```
```suggestion
			return errors.New("status won't be updated because object is marked for deletion")
```

qbarrand commented on internal/controllers/state/statusupdater.go:
```diff
@@ -31,40 +53,71 @@ func NewStatusUpdater(kubeClient clients.ClientsInterface) StatusUpdater {
+		err := su.kubeClient.Get(ctx, types.NamespacedName{Name: sr.Name, Namespace: ""}, sr)
 		if apierrors.IsNotFound(err) {
-			return
+			return errors.New("Could not update status because the object does not exist")
```
```suggestion
			return errors.New("could not update status because the object does not exist")
```

qbarrand commented on internal/controllers/state/statusupdater.go:
```diff
@@ -31,40 +53,71 @@ func NewStatusUpdater(kubeClient clients.ClientsInterface) StatusUpdater {
-func (su *statusUpdater) UpdateWithState(ctx context.Context, sr *v1beta1.SpecialResource, state string) {
+// SetAsProgressing changes SpecialResource's Progressing condition as true and changes Ready and Errored conditions to false, and updates the status in the API.
+func (su *statusUpdater) SetAsProgressing(ctx context.Context, sr *v1beta1.SpecialResource, reason, message string) error {
+	return su.updateWithMutator(ctx, sr, func(o *v1beta1.SpecialResource) {
```
Why have a `updateWithMutator`? Could we just set the status and then use a [`StatusWriter`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client#StatusWriter) to save the SpecialResource in Kubernetes?

pmtk replied:
I'll investigate it, thanks for the idea

pmtk replied:
Please take a look

qbarrand commented on internal/controllers/state/statusupdater.go:
```diff
@@ -31,40 +53,71 @@ func NewStatusUpdater(kubeClient clients.ClientsInterface) StatusUpdater {
+	}
+
+	mutator(&update)
 	update.DeepCopyInto(sr)
```
I'm not sure why we copy the object here. Any idea? Could we remove that convoluted logic?

yevgeny-shnaidman replied:
i agree with @qbarrand , i think that those functions should be more straightforward:
1) no need to check for CR presence - in case it is missing, update will fail and we go for another loop where we will check the presence of the CR
2) SetAs.. function should return error (maybe not SetAsError) and reconciliation loop should fail in case CR is not updated
3) no need for deep copy, we can work on the current CR

pmtk replied:
Thanks for the comments, I'll check it out

pmtk replied:
btw. I think that the flow here (Get + Update) is defensive in case of CR change in API (and then copy the freshest object back into what we're handling). I'll watch out for this when investigating simplification of this function, but it might be overkill for just updating the status (and not specs)

pmtk replied:
Made changes as suggested. Had to change some `SpecialResource` params or struct fields to pointers.

yevgeny-shnaidman commented on controllers/specialresource.go:
```diff
@@ -56,7 +56,7 @@ func SpecialResourcesReconcile(ctx context.Context, r *SpecialResourceReconciler
 		}
 		parent, err := r.Storage.CheckConfigMapEntry(ctx, req.Name, obj)
 		if err != nil {
-			r.StatusUpdater.UpdateWithState(ctx, &r.parent, fmt.Sprintf("%v", err))
```
don't we want to set the condition/ state here to error?

pmtk replied:
This is a case when the SR was not found, so what would we update?
And I'm afraid that `r.parent` might just contain something from previous reconciliation.

yevgeny-shnaidman commented:
@pmtk can we have an explanation section (maybe in the PR or in some other doc) that documents exactly what we have agreed on regarding conditions/states and when they are set?

pmtk commented:
@yevgeny-shnaidman  I've added a detailed description to the PR's desc.
I'm not particularly happy about quick changes between `Errored` and `Progressing`, but in current implementation of SRO the reconciliation will be tried endlessly (there's no situation of SRO stopping retries requiring user's intervention).
However, I feel like the `Errored` condition is important part of the UX and it might be useful if we introduce SR without helm charts to clearly communicate issues as requiring user to read the logs is not ideal.

pmtk committed:
Simplify StatusUpdater

StatusUpdater will no longer GET newest version of an object from API.
It'll operate directly on an object passed in.
This force change of some fields of SpecialResourceReconciler and
function parameters from values to pointers.

pmtk committed:
Propagate errors from SetAs{Ready,Progressing}

Errors from SetAsErrored are only printed out.

qbarrand commented on api/v1beta1/specialresource_types.go:
```diff
@@ -137,12 +137,27 @@ type SpecialResourceDependency struct {
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```
I'm sorry for realizing this late, but can we modify the API object without bumping is version?
I don't think we generally can, although in this case leaving the `State` field might make this new API backwards compatible.

pmtk replied:
No worries.
Since this is part of `.Status` and not `.Spec` I don't think we'd break anything. `.Status` is not something user should create when applying CR. 
Unless there's some other Operator programmatically watching CR's `.Status` before doing its own thing? I'd argue that that the `.Status.State` wasn't particularly good object to observe ;)

We can leave it (it sounds sensibly and responsibly), but on the other hand now is the best time to make (potentially) breaking changes (the later in upstreaming process the harder it'll be).
How about removing the code (leave it empty at all times) or leaving the code (then what will be the best moment to deprecate and remove this)?

pmtk replied:
Brought `State` back to life

qbarrand replied:
Thank you!

pmtk committed:
Bring back State field for API compat

pmtk committed:
Fix after rebase

qbarrand commented on controllers/specialresource.go:
```diff
@@ -57,7 +58,7 @@ func SpecialResourcesReconcile(ctx context.Context, r *SpecialResourceReconciler
 		parent, err := r.Storage.CheckConfigMapEntry(ctx, req.Name, obj)
 		if err != nil {
-			r.StatusUpdater.UpdateWithState(ctx, &r.parent, fmt.Sprintf("%v", err))
+			log.Error(err, "")
```
I think this should have an error message.

pmtk replied:
You're right, added

pmtk committed:
Add error msg

qbarrand commented:
/lgtm

bthurber commented:
/approve
